src/gt4py/next/iterator/type_system/type_synthesizer.py

```diff
@@ -270,82 +276,172 @@ def apply_lift(
     return apply_lift
 
 
-def _convert_as_fieldop_input_to_iterator(
-    domain: it_ts.DomainType, input_: ts.TypeSpec
-) -> it_ts.IteratorType:
-    # get the dimensions of all non-zero-dimensional field inputs and check they agree
+def _collect_and_check_dimensions(input_: ts.TypeSpec) -> list[common.Dimension]:
+    """
+    Extracts dimensions from non-zero-dimensional field inputs and ensures they match.
+    """
     all_input_dims = (
         type_info.primitive_constituents(input_)
         .if_isinstance(ts.FieldType)
         .getattr("dims")
         .filter(lambda dims: len(dims) > 0)
         .to_list()
     )
-    input_dims: list[common.Dimension]
-    if all_input_dims:
-        assert all(cur_input_dims == all_input_dims[0] for cur_input_dims in all_input_dims)
-        input_dims = all_input_dims[0]
+    if not all_input_dims:
+        return []
     else:
-        input_dims = []
-
-    element_type: ts.DataType
-    element_type = type_info.apply_to_primitive_constituents(type_info.extract_dtype, input_)
-
-    # handle neighbor / sparse input fields
-    defined_dims = []
-    neighbor_dim = None
-    for dim in input_dims:
-        if dim.kind == common.DimensionKind.LOCAL:
-            assert neighbor_dim is None
-            neighbor_dim = dim
-        else:
-            defined_dims.append(dim)
-    if neighbor_dim:
-        element_type = ts.ListType(element_type=element_type, offset_type=neighbor_dim)
+        assert all(cur_input_dims == all_input_dims[0] for cur_input_dims in all_input_dims)
+        return all_input_dims[0]
```

Comment:
Why has the order of the branches changed here? I think in general makes more sense to express the condition in positive, without the `not`. Additionally, please don't add an `else` after a unconditional return, is not needed and slightly misleading.